app/frontend/wallet/helpers/CachedDeriveXpubFactory.ts

```diff
@@ -28,14 +37,55 @@ function CachedDeriveXpubFactory(derivationScheme, deriveXpubHardenedFn) {
     return await derivedXpubs[memoKey]
   }
 
-  async function deriveXpubNonhardenedFn(derivationPath) {
+  async function deriveXpubNonhardenedFn(derivationPath: BIP32Path) {
     const lastIndex = derivationPath.slice(-1)[0]
     const parentXpub = await deriveXpub(derivationPath.slice(0, -1))
-
     return deriveChildXpub(parentXpub, lastIndex, derivationScheme.ed25519Mode)
   }
 
-  return deriveXpub
+  function createPathBulk(derivationPath: BIP32Path): BIP32Path[] {
+    const paths: BIP32Path[] = []
+    const accountIndex = derivationPath[2] - HARDENED_THRESHOLD
+    const currentAccountPage = Math.floor(accountIndex / MAX_BULK_EXPORT_AMOUNT)
+
+    /*
+    * in case of the account 0 we append also the byron path
+    * since during byron era only the first account was used
+    */
+    if (accountIndex === 0) paths.push(BYRON_V2_PATH)
+
+    for (let i = 0; i < MAX_BULK_EXPORT_AMOUNT; i += 1) {
+      const nextAccountIndex = currentAccountPage * MAX_BULK_EXPORT_AMOUNT + i + HARDENED_THRESHOLD
+      const nextAccountPath = [...derivationPath.slice(0, -1), nextAccountIndex]
+      paths.push(nextAccountPath)
+    }
+    return paths
+  }
+
+  async function deriveXpubHardenedFn(derivationPath: BIP32Path): Promise<any> {
+    const paths = shouldExportPubKeyBulk ? createPathBulk(derivationPath) : [derivationPath]
+    const xPubBulk = await deriveXpubFn(paths)
+    const _derivedXpubs = {}
+    xPubBulk.forEach((xpub: Buffer, i: number) => {
+      const memoKey = JSON.stringify(paths[i])
+      _derivedXpubs[memoKey] = Promise.resolve(xpub)
+    })
+    return _derivedXpubs
+  }
+
+  function cleanXpubCache() {
+    const _derivedXpubs = {}
+    Object.entries(derivedXpubs).map(([key, xpubPromise]: [string, Promise<Buffer>]) => {
+      xpubPromise
+        .then((xpub) => {
+          _derivedXpubs[key] = Promise.resolve(xpub)
+        })
+        .catch((e) => null)
+    })
+    derivedXpubs = _derivedXpubs
+  }
+
+  return {deriveXpub, cleanXpubCache}
```

Comment:
cleanXpubCache seems like a very low-level thing to expose. I'm actually trying to understand why it's needed (i.e. what could happen if it wouldn't exist) - could you please explain? And isn't there a way to keep the cache cleanup contained within CachedDeriveXpubFactory ?

Reply:
Afaik there is no way to clean the xpubCache inside CachedDeriveXpubFactory if we are deriving the keys asynchronously. I agree it seems a bit low level. Only way afaik would be to drop the async behaviour. 

To answer your question, we need to clear the cache in case a user (lets say) declines the pubkey export prompt. If we didn't clean it he would need to reload and try again.